src/fetch/payouts.py

```diff
@@ -0,0 +1,340 @@
+        assert secondary_reward_cow >= 0, "invalid secondary_reward_cow"
+
+        self.solver = solver
+        self.solver_name = solver_name
+        self.reward_target = reward_target
+        self.exec_cost = exec_cost
+        self.payment_eth = payment_eth
+        self.slippage_eth = slippage_eth
+        self.reward_cow = reward_cow
+        self.secondary_reward_eth = secondary_reward_eth
+        self.secondary_reward_cow = secondary_reward_cow
+
+    @classmethod
+    def from_series(cls, frame: Series) -> RewardAndPenaltyDatum:
+        """Constructor from row in Dataframe"""
+        slippage = (
+            int(frame["eth_slippage_wei"])
+            if not math.isnan(frame["eth_slippage_wei"])
+            else 0
+        )
+        solver = frame["solver"]
+        reward_target = frame["reward_target"]
+        if reward_target is None:
+            logging.warning(f"solver {solver} without reward_target. Using solver")
+            reward_target = solver
+
+        return cls(
+            solver=Address(solver),
+            solver_name=frame["solver_name"],
+            reward_target=Address(reward_target),
+            payment_eth=int(frame["payment_eth"]),
+            slippage_eth=slippage,
+            reward_cow=int(frame["reward_cow"]),
+            exec_cost=int(frame["execution_cost_eth"]),
+            secondary_reward_eth=int(frame["secondary_reward_eth"]),
+            secondary_reward_cow=int(frame["secondary_reward_cow"]),
+        )
+
+    def total_outgoing_eth(self) -> int:
+        """Total outgoing amount (including slippage) for the payout."""
+        return self.payment_eth + self.secondary_reward_eth + self.slippage_eth
+
+    def total_cow_reward(self) -> int:
+        """Total outgoing COW token reward"""
+        return self.reward_cow + self.secondary_reward_cow
+
+    def total_eth_reward(self) -> int:
+        """Total outgoing COW token reward"""
+        return self.payment_eth - self.exec_cost + self.secondary_reward_eth
+
+    def is_overdraft(self) -> bool:
+        """
+        True if the solver's complete combined data results in a net negative
+        """
+        return self.total_outgoing_eth() < 0
+
+    def as_payouts(self) -> list[Transfer]:
+        """
+        Isolating the logic of how solvers are paid out according to their
+            execution costs, rewards and slippage
+        """
+        # TODO - decide how to approach the "order of operations" regarding slippage and rewards
+        #  Seemingly, the simplest option is to deal with Reward Scheme First and Slippage after.
+        #  However, there is this issue with the fact that payment can be negative
+        #  (which didn't exist before)
+        # We do not handle overdraft scenario here!
+        assert not self.is_overdraft()
+        reimbursement_eth = int(self.exec_cost + self.slippage_eth)
+        # TODO: the following value is not available in COW at the moment
+        reimbursement_cow = int(0)
```

Comment:
@fhenneke - so it turns out we cannot just have this value ready from the data frame because it is based on a value that is only ever in ETH. This means the class containing the payout details will also have to have an instance of token converter (so that we can convert the eth value above here whenever required)

Reply:
I have decided to go with the following approach:

```
        reward_eth = int(self.total_eth_reward())
        reward_cow = int(self.total_cow_reward())

        reimbursement_eth = int(self.exec_cost + self.slippage_eth)
        # We do not have access to token conversion here, but we do have other converted values
        # x_eth:x_cow = y_eth:y_cow --> y_cow = y_eth * x_cow / x_eth
        reimbursement_cow = (reimbursement_eth * reward_cow) // reward_eth
```

Please let me know what we think about this.

Reply:
nice solution, sounds good!